### DISCUSSION TITLE
How to have an exhaustive check of top level states to render come component ?

### DISCUSSION THREAD

**Discussion by stevebeauge** [statelyai/xstate#4032] [Q&A]
I'm new to xstate, I hope my question is not obvious.

I'm implementing a "wizard" using xstate + react + typescript.

Basically, I have multiple UI steps, where each step may depend on previous steps input. Each step may either "go to next step" or submit currently known data.

I designed my flow chart to have one "initialization" state which retrieve contextual data and one state for each step.

Some states are basic states but some are nested state (because of async operation that should disable the forms while operating).

My top level component (where the useMachine hook is called) should render the UI depending on the top level state. Nested states should be implementation details of each step's UI.

To achieve this kind of "routing", I have a cascade of if/else:

````typescript
if(machine.matches('InitializeState')) {
    return <Spinner/>;
} else if(machine.matches('Step1')) {
    return <Step1/>;
} else if(machine.matches('Step2')) {
    return <Step2/>;
} else if(machine.matches('Confim')) {
    return <Confirm/>;
} else {
   throw new Error(`State ${machine.value} not implemented`);
}
````

This is working, but will highlight missing state only at runtime.

I often use TS exhaustive type check in `switch` statements. But is there a way to apply this to xstate machine ?

I can "extract" the top level state by splitting the state before the first dot, but I have no type (at least not found) to compare to

**Comment by davidkpiano** [comment]
Great question. Since states can be nested (e.g. `state.matches({ parent: 'child' })`), normal exhaustive matching isn't advanced enough to handle those cases yet, to the best of my knowledge. However, this is a great idea and we'll experiment with this to see if we can strongly infer the types of state values, and automatically enable exhaustive matching in XState.

**Comment by chaitanyaj2222** [comment]
@stevebeauge, To consider only `top-level` state in your _switch_ or _if-else_ statement conditions, you can add a `tag` to the top-level states of your machine & use `machine.hasTag('tagName')` instead of using `machine.matches('stateValue')` to check if the machine is currently inside of top-level state. Please refer to this [section](https://xstate.js.org/docs/guides/statenodes.html#tags) of the Xstate documentation on how to use tags.
